compiler/rustc_typeck/src/coherence/orphan.rs

```diff
@@ -277,42 +277,78 @@ fn emit_orphan_check_error<'tcx>(
         }
         traits::OrphanCheckErr::UncoveredTy(param_ty, local_type) => {
             let mut sp = sp;
-            for param in generics.params {
-                if param.name.ident().to_string() == param_ty.to_string() {
-                    sp = param.span;
+
+            match *param_ty.kind() {
+                ty::Projection(..) => {
+                    sp = self_ty_span;
                 }
-            }
+                _ => {
+                    for param in generics.params {
+                        if param.name.ident().to_string() == param_ty.to_string() {
+                            sp = param.span;
+                        }
+                    }
+                }
+            };
 
             match local_type {
-                Some(local_type) => struct_span_err!(
-                    tcx.sess,
-                    sp,
-                    E0210,
-                    "type parameter `{}` must be covered by another type \
-                    when it appears before the first local type (`{}`)",
-                    param_ty,
-                    local_type
-                )
-                .span_label(
-                    sp,
-                    format!(
-                        "type parameter `{}` must be covered by another type \
-                    when it appears before the first local type (`{}`)",
-                        param_ty, local_type
-                    ),
-                )
-                .note(
-                    "implementing a foreign trait is only possible if at \
+                Some(local_type) => {
+                    let mut err = match *param_ty.kind() {
+                        ty::Projection(..) => {
+                            let mut err = struct_span_err!(
+                                tcx.sess,
+                                sp,
+                                E0210,
+                                "type projection `{}` must be covered by another type \
```

Comment:
"associated type" might be better name for it, since "projection" is a bit of compiler language